FEATURE: Ansible module for getting box info

caladd opened the pull request:
An Ansible module for returning box info from the Stacki database. The info returned is the same as `list box`.

The module takes `name` as an optional parameter to filter the returned data by box name. The data returned is a list of `boxes`.

Example playbook:
```
---
- hosts: localhost
  tasks:
    - name: Get info on all the boxes
      stacki_box_info:
      register: result

    - name: Multiple boxes output
      debug:
        var: result

    - name: Get the frontend box info
      stacki_box_info:
        name: frontend
      register: result

    - name: Single box output
      debug:
        var: result
```

Output of the debug commands, showing the structure of the data returned:
```
TASK [Multiple boxes output] ***********************************
ok: [localhost] => {
    "result": {
        "boxes": [
            {
                "carts": [
                    "vagrant"
                ],
                "name": "default",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            },
            {
                "carts": [],
                "name": "frontend",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            }
        ],
        "changed": false,
        "failed": false
    }
}

TASK [Single box output] ***********************************
ok: [localhost] => {
    "result": {
        "boxes": [
            {
                "carts": [],
                "name": "frontend",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            }
        ],
        "changed": false,
        "failed": false
    }
}
```

bsanders commented on common/src/stack/ansible/plugins/module_utils/stacki.py:
```diff
@@ -0,0 +1,50 @@
+	"""Runs a stack command, returning StackCommandError if there is an error."""
+
+	# Create our command to run
+	command_args = ["/opt/stack/bin/stack"]
```
This effectively ties the ansible modules to running on the frontend, no?

caladd replied:
Correct, which I decided was appropriate for the first version/MVP. Since our Ansible playbooks will be running on the frontend anyways, and the Ansible modules only installed on the frontend during the normal workflow, it should satisfy SNAPI's needs.

In a version 2, especially if we change our Webservice to be JWT based, it might make sense to do the work to have this function use the webservice and deal with authentication and whatnot. If we switch to using JWT, we would only need to supply an API endpoint and the JWT refresh token in any playbook.

But, for the time being is simpler to just assume the modules are running on the frontend.

bsanders replied:
Also, maybe we already talked about this and I forgot, but is there a reason to shell out to `stack` rather than `import stack` and handle it all via Python?

caladd replied:
Part of it for simplicity reasons, I'm not sure how loading all of Stacki's python stuff into Ansible's python process would go. Also, I'm concerned about the Stacki DB cache layer, we saw lots of bugs with Apache not shelling out to run Stacki commands.

What are the reasons to *not* shell out, besides the overhead of spinning up a process?

langkurt commented on common/src/stack/ansible/plugins/modules/stacki_box_info.py:
```diff
@@ -0,0 +1,122 @@
+
+	try:
+		for box in run_stack_command("list.box", args):
+			# Split pallets, carts, and repos into lists
```
Shouldnt `os` be part of the result as well?

caladd replied:
It is part of the result, already in the `box` dict. Only the pallets, carts, and repos have to be post-processed from what comes out of `list.box`, to convert them to lists. They come out as a space separated string, which is harder to work with in an Ansible playbook.

caladd committed:
FEATURE: Ansible module for getting box info

An Ansible module for returning box info from the Stacki database. The info returned is the same as `list box`.

The module takes `name` as an optional parameter to filter the returned data by box name. The data returned is a list of `boxes`.

Example playbook:
```
---
- hosts: localhost
  tasks:
    - name: Get info on all the boxes
      stacki_box_info:
      register: result

    - name: Multiple boxes output
      debug:
        var: result

    - name: Get the frontend box info
      stacki_box_info:
        name: frontend
      register: result

    - name: Single box output
      debug:
        var: result
```

Output of the debug commands, showing the structure of the data returned:
```
TASK [Multiple boxes output] ***********************************
ok: [localhost] => {
    "result": {
        "boxes": [
            {
                "carts": [
                    "vagrant"
                ],
                "name": "default",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            },
            {
                "carts": [],
                "name": "frontend",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            }
        ],
        "changed": false,
        "failed": false
    }
}

TASK [Single box output] ***********************************
ok: [localhost] => {
    "result": {
        "boxes": [
            {
                "carts": [],
                "name": "frontend",
                "os": "sles",
                "pallets": [
                    "stacki-5.6rc1_20200724_24b0c82-sles12",
                    "SLES-12sp3-sles12"
                ],
                "repos": []
            }
        ],
        "changed": false,
        "failed": false
    }
}
```

bsanders commented:
Rebased and merged
